clib/helpstr.c

```diff
@@ -43,115 +125,107 @@ int *ierr;
-    decloc = strrchr(outbuf,'.');
-    if(declocstr !=NULL) {
-      strcpy(runstr,outbuf);
-      *ierr = 0;
-      break;
-    } else if(decloc != NULL) {
-      ch1=*(decloc+1);
-      ch2=*(decloc+2);
-      ch3=*(decloc+3);
-      if((ch1== '_' ||
-	  (ch1 == '3' &&  K4K3  == *rack) ||
-	  (ch1 == 'm' &&  MK3   == *rack) ||
-	  (ch1 == 'n' && (MK3   == *rack || MK4   == *rack ||
-	                  LBA4  == *rack)) ||
-	  (ch1 == 'e' && (MK3   == *rack || MK4   == *rack ||
-			  VLBA  == *rack || VLBA4 == *rack ||
-	                  LBA4  == *rack || DBBC  == *rack)) ||
-	  (ch1 == 'f' && (MK3   == *rack || MK4   == *rack ||
-			  K4K3  == *rack || K4MK4 == *rack ||
-			  K4    == *rack || LBA4  == *rack)) ||
-	  (ch1 == '4' &&  MK4   == *rack) ||
-	  (ch1 == 's' &&  S2    == *rack) ||
-	  (ch1 == 'g' && (MK4   == *rack || VLBA  == *rack ||
-			  VLBA4 == *rack || K4MK4 == *rack ||
-			  LBA4  == *rack)) ||
-	  (ch1 == 'h' && (MK4   == *rack || VLBA4 == *rack ||
-			  K4MK4 == *rack || LBA4  == *rack)) ||
-	  (ch1 == 'i' && (MK4   == *rack || VLBA  == *rack ||
-			  VLBA4 == *rack || K4MK4 == *rack)) ||
-	  (ch1 == 'v' && (VLBA  == *rack)) ||
-	  (ch1 == 'w' && (VLBA  == *rack || VLBA4 == *rack)) ||
-	  (ch1 == 'k' && (K4K3  == *rack || K4MK4 == *rack ||
-			  K4    == *rack)) ||
-	  (ch1 == 'l' && (LBA   == *rack || LBA4 == *rack)) ||
-	  (ch1 == 'd' &&  DBBC  == *rack) ||
-	  (ch1 == 'a' &&  0     != *rack)) &&
-	 (ch2 == '_' || ch2 == '+' ||
-	  (ch2 == 'k' &&  K4    == *drive1) ||
-	  (ch2 == 'm' &&  MK3   == *drive1) ||
-	  (ch2 == 'n' && (MK3   == *drive1 || MK4 == *drive1)) ||
-	  (ch2 == '4' &&  MK4   == *drive1) ||
-	  (ch2 == 's' &&  S2    == *drive1) ||
-	  (ch2 == 'w' && (VLBA  == *drive1 || VLBA4 == *drive1)) ||
-	  (ch2 == 'a' &&  0     != *drive1) ||
-	  (ch2 == 'l' && (MK3   == *drive1 || MK4   == *drive1 ||
-			  VLBA  == *drive1 || VLBA4 == *drive1 ))) &&
-	 (ch3 == '_' || ch3 == '+' ||
-	  (ch3 == 'k' &&  K4    == *drive2) ||
-	  (ch3 == 'm' &&  MK3   == *drive2) ||
-	  (ch3 == 'n' && (MK3   == *drive2 || MK4 == *drive2)) ||
-	  (ch3 == '4' &&  MK4   == *drive2) ||
-	  (ch3 == 's' &&  S2    == *drive2) ||
-	  (ch3 == 'w' && (VLBA  == *drive2 || VLBA4 == *drive2)) ||
-	  (ch3 == 'a' &&  0     != *drive2) ||
-	  (ch3 == 'l' && (MK3   == *drive2 || MK4   == *drive2 ||
-			  VLBA  == *drive2 || VLBA4 == *drive2 )))
-	 &&
-	 ((*drive1 !=0 && *drive2 !=0 && ((ch2 == '+' || ch3 == '+') ||
-					  (ch2 =='_' && ch3 == '_'))) ||
-	  ((*drive1 ==0 || *drive2 == 0) && (ch2 != '+' && ch3 != '+')))
-	 ) {
-        strcpy(runstr,outbuf);
-        *ierr = 0;
-        break;
-      }
-    }
-  }
-  fclose(idum);
+	char name[MAX_STRING] = {};
+	char *decloc;
+	char ch1, ch2, ch3;
+
+	*ierr = -3;
+
+	if (*clength > MAX_STRING) {
+		*ierr = -2;
+		return;
+	}
+
+	strncpy(name, cnam, *clength);
+	name[*clength] = '\0';
+
+	decloc = strchr(name, '.');
+	if (decloc != NULL) {
+		char *p = check_help_dirs(name);
```

Comment:
I think this actually does more than advertised. Eg giving it the argument `../../etc/passwd`, it will return successfully. It might be easier to restore the method the old code used, of appending "." if not present

Reply:
Sort of symmetric with other case I guess. I think you are saying that it could branch out of the directory and display some other file entirely. That seems like it would be safe, but might have some unpleasant effect if the file is binary. Maybe it should be rejected if the first character is dot and if the first dot isn't the last minus third(?) character (making it conform to the expected format). I suppose the old code had this defect as well. Or have I gone down a rabbit hole. :)

Reply:
I was more thinking down-the-line we might like to expose this function to a user who shouldn't be able to list any file. 

The prefix case is safe from directory traversal, as it lists the directory contents of the help directories directly. That's why I wanted to reduce this case to the prefix one.

Reply:
Not quite sure what you are getting at about down the line, but I made a mistake in the call to access, it should be:
````
		if ( ! access(path, R_OK)) {
			return path;
		}
````
Sorry about that. With that, it rejects **help=../../etc/passwd**, but branches out for **help=../../../etc/passwd**, which of course is a problem, back to my first reply. I can look at that later if you haven't.

Reply:
Fixed the inverted logic. I went with just adding a check at the top for a "/" in the name and aborting.